pytorch_lightning/core/step_result.py

```diff
@@ -11,603 +11,534 @@
-        # ensure tensors are detached
-        if isinstance(val, torch.Tensor) and key != 'minimize':
-            val = val.detach()
-        self[key] = val
+    @property
+    def forked(self) -> bool:
+        return self.on_step and self.on_epoch
 
-    def __getstate__(self):
-        return self
+    @property
+    def forked_step_name(self) -> str:
+        if self.forked:
+            return self.name + "_step"
+        return self.name
 
-    def __setstate__(self, d):
-        self.update(d)
+    @property
+    def forked_epoch_name(self) -> str:
+        if self.forked:
+            return self.name + "_epoch"
+        return self.name
 
     @property
-    def minimize(self) -> Optional[Tensor]:
-        return self.get('minimize', None)
+    def is_tensor_and_mean_reduction(self) -> bool:
+        return self.is_tensor and self.reduce_fx == torch.mean
 
-    @minimize.setter
-    def minimize(self, val: Optional[torch.Tensor]) -> None:
-        if val is not None:
-            if not isinstance(val, Tensor):
-                raise ValueError(f"`Result.minimize` must be a `torch.Tensor`, found: {val}")
-            if val.grad_fn is None:
-                raise RuntimeError("`Result.minimize` must have a `grad_fn`")
-        self['minimize'] = val
+    @property
+    def is_tensor_and_max_reduction(self) -> bool:
+        return self.is_tensor and (self.reduce_fx in (torch.max, max))
 
-    def log(
-        self,
-        name: str,
-        value: Any,
-        prog_bar: bool = False,
-        logger: bool = True,
-        on_step: bool = False,
-        on_epoch: bool = True,
-        reduce_fx: Callable = torch.mean,
-        enable_graph: bool = False,
-        dataloader_idx: Optional[int] = None,
-    ):
-        # no metrics should be logged with graphs
-        if not enable_graph and isinstance(value, torch.Tensor):
-            value = value.detach()
+    @property
+    def is_tensor_and_min_reduction(self) -> bool:
+        return self.is_tensor and (self.reduce_fx in (torch.min, min))
 
-        if isinstance(value, torch.Tensor) and value.device.type == "xla":
-            value = value.cpu()
 
-        if 'meta' not in self:
-            self.__setitem__('meta', {})
-
-        # if user requests both step and epoch, then we split the metric in two automatically
-        # one will be logged per step. the other per epoch
-        was_forked = False
-        if on_step and on_epoch:
-            was_forked = True
-
-            # set step version
-            step_name = f'{name}_step'
-
-            self.__set_meta(
-                step_name,
-                value,
-                prog_bar,
-                logger,
-                on_step=True,
-                on_epoch=False,
-                reduce_fx=reduce_fx,
-                forked=False,
-                dataloader_idx=dataloader_idx,
-            )
+class ResultMetric(Metric):
 
-            self.__setitem__(step_name, value)
+    DTYPE = torch.float32
```

Comment:
Instead of having this, maybe inherit from `Metric` and `DeviceDtypeModuleMixin`?

This would add `self.dtype` and `self.device` and all the logic to carry changes like converting it to half or something similar.

Reply:
Done !